File: src/main/java/me/ag2s/umdlib/tool/StreamReader.java
package me.ag2s.umdlib.tool;

import java.io.IOException;
import java.io.InputStream;

public class StreamReader {
   private InputStream is;
   private long offset;
   private long size;

   public long getOffset() {
      return this.offset;
   }

   public void setOffset(long offset) {
      this.offset = offset;
   }

   public long getSize() {
      return this.size;
   }

   public void setSize(long size) {
      this.size = size;
   }

   private void incCount(int value) {
      int temp = (int)(this.offset + (long)value);
      if (temp < 0) {
         temp = Integer.MAX_VALUE;
      }

      this.offset = (long)temp;
   }

   public StreamReader(InputStream inputStream) throws IOException {
      this.is = inputStream;
   }

   public short readUint8() throws IOException {
      byte[] b = new byte[1];
      this.is.read(b);
      this.incCount(1);
      return (short)(b[0] & 255);
   }

   public byte readByte() throws IOException {
      byte[] b = new byte[1];
      this.is.read(b);
      this.incCount(1);
      return b[0];
   }

   public byte[] readBytes(int len) throws IOException {
      if (len < 1) {
         System.out.println(len);
         throw new IllegalArgumentException("Length must > 0: " + len);
      } else {
         byte[] b = new byte[len];
         this.is.read(b);
         this.incCount(len);
         return b;
      }
   }

   public String readHex(int len) throws IOException {
      if (len < 1) {
         System.out.println(len);
         throw new IllegalArgumentException("Length must > 0: " + len);
      } else {
         byte[] b = new byte[len];
         this.is.read(b);
         this.incCount(len);
         return UmdUtils.toHex(b);
      }
   }

   public short readShort() throws IOException {
      byte[] b = new byte[2];
      this.is.read(b);
      this.incCount(2);
      short x = (short)((b[0] & 255) << 8 | (b[1] & 255) << 0);
      return x;
   }

   public short readShortLe() throws IOException {
      byte[] b = new byte[2];
      this.is.read(b);
      this.incCount(2);
      short x = (short)((b[1] & 255) << 8 | (b[0] & 255) << 0);
      return x;
   }

   public int readInt() throws IOException {
      byte[] b = new byte[4];
      this.is.read(b);
      this.incCount(4);
      int x = (b[0] & 255) << 24 | (b[1] & 255) << 16 | (b[2] & 255) << 8 | (b[3] & 255) << 0;
      return x;
   }

   public int readIntLe() throws IOException {
      byte[] b = new byte[4];
      this.is.read(b);
      this.incCount(4);
      int x = (b[3] & 255) << 24 | (b[2] & 255) << 16 | (b[1] & 255) << 8 | (b[0] & 255) << 0;
      return x;
   }

   public void skip(int len) throws IOException {
      this.readBytes(len);
   }

   public byte[] read(byte[] b) throws IOException {
      this.is.read(b);
      this.incCount(b.length);
      return b;
   }

   public byte[] read(byte[] b, int off, int len) throws IOException {
      this.is.read(b, off, len);
      this.incCount(len);
      return b;
   }
}

File: src/main/kotlin/me/ag2s/umdlib/tool/StreamReader.kt
package me.ag2s.umdlib.tool

import java.io.InputStream

class StreamReader(private val inputStream: InputStream) {
    var offset: Long = 0
    var size: Long = 0

    private fun incCount(value: Int) { offset = (offset + value).let { if (it < 0) Int.MAX_VALUE.toLong() else it } }
    fun readUint8(): Short = ByteArray(1).also { inputStream.read(it); incCount(1) }[0].let { (it.toInt() and 0xff).toShort() }
    fun readByte(): Byte = ByteArray(1).also { inputStream.read(it); incCount(1) }[0]
    fun readBytes(length: Int): ByteArray { require(length >= 1) { "Length must > 0: $length" }; return ByteArray(length).also { inputStream.read(it); incCount(length) } }
    fun readHex(length: Int): String = UmdUtils.toHex(readBytes(length))
    fun readShort(): Short { val b=readBytes(2); return (((b[0].toInt() and 0xff) shl 8) or (b[1].toInt() and 0xff)).toShort() }
    fun readShortLe(): Short { val b=readBytes(2); return (((b[1].toInt() and 0xff) shl 8) or (b[0].toInt() and 0xff)).toShort() }
    fun readInt(): Int { val b=readBytes(4); return (b[0].toInt() and 0xff shl 24) or (b[1].toInt() and 0xff shl 16) or (b[2].toInt() and 0xff shl 8) or (b[3].toInt() and 0xff) }
    fun readIntLe(): Int { val b=readBytes(4); return (b[3].toInt() and 0xff shl 24) or (b[2].toInt() and 0xff shl 16) or (b[1].toInt() and 0xff shl 8) or (b[0].toInt() and 0xff) }
    fun skip(length: Int) { readBytes(length) }
    fun read(bytes: ByteArray): ByteArray { inputStream.read(bytes); incCount(bytes.size); return bytes }
    fun read(bytes: ByteArray, off: Int, len: Int): ByteArray { inputStream.read(bytes,off,len); incCount(len); return bytes }
}
